### PR TITLE
Feature 3.5/sharding config

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Documentation companion PR: https://github.com/arangodb/docs/pull/181
Enterprise companion PR: https://github.com/arangodb/enterprise/pull/331

* Added startup option `--cluster.max-number-of-shards` for restricting the
  maximum number of shards when creating new collections. The default
  value for this setting is `1000`, which is also the previously hard-coded
  built-in limit. A value of `0` for this option means "unrestricted".
  When the setting is adjusted, it will not affect already existing
  collections, but only collections that are created or restored
  afterwards.

* Added startup options for managing the replication factor for newly 
  created collections:

  - `--cluster.min-replication-factor`: this settings controls the minimum
    replication factor value that is permitted when creating new collections.
    No collections can be created which have a replication factor value 
    below this setting's value. The default value is 1. 
  - `--cluster.max-replication-factor`: this settings controls the maximum
    replication factor value that is permitted when creating new collections.
    No collections can be created which have a replication factor value 
    above this setting's value. The default value is 10.
  - `--cluster.default-replication-factor`: this settings controls the default
    replication factor value that is used when creating new collections and 
    no value of replication factor has been specified. 
    If no value is set for this option, the value of the option
    `--cluster.min-replication-factor` will be used. 

- [ ] Bug-Fix for *devel-branch* (i.e. no need for backports?)
- [ ] Bug-Fix for a *released version* (did you remember to port this to all relevant release branches?)
- [x] Strictly *new functionality* (i.e. a new feature / new option, no need for porting)
- [x] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)
- [x] The behavior change can be verified via automatic tests

### Testing & Verification

This PR adds tests that were used to verify all changes:

- [x] Added new **integration tests** (shell_client, shell_server)

https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/6488/

### CODE REVIEW

**Comment by jsteemann** [comment]
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/6492/

File: arangod/Cluster/ClusterFeature.h [hunk 1]
```diff
@@ -66,7 +66,11 @@ class ClusterFeature : public application_features::ApplicationFeature {
   std::string _myRole;
   std::string _myEndpoint;
   std::string _myAdvertisedEndpoint;
-  uint32_t _systemReplicationFactor = 2;
+  std::uint32_t _defaultReplicationFactor = 0; // a value of 0 means it will use the min replication factor 
+  std::uint32_t _systemReplicationFactor = 2;
+  std::uint32_t _minReplicationFactor = 1;     // minimum replication factor (0 = unrestricted)
```
Comment (mchacki) [inline arangod/Cluster/ClusterFeature.h]:
```suggestion
  std::uint32_t _minReplicationFactor = 1; 
```

1 is already unrestricted, you cannot create any collection lower than this.

File: js/apps/system/_admin/aardvark/APP/frontend/js/views/collectionsView.js [hunk 2]
```diff
@@ -537,6 +549,29 @@
                 false
               )
             );
+            tableContent.push(
```
Comment (mchacki) [inline js/apps/system/_admin/aardvark/APP/frontend/js/views/collectionsView.js]: I like this being moved to a more visible place

File: tests/js/common/shell/shell-cluster-collection.js [hunk 3]
```diff
@@ -411,11 +412,11 @@ function ClusterCollectionSuite () {
 /// @brief test create
 ////////////////////////////////////////////////////////////////////////////////
 
-    testCreateInvalidShardKeys1 : function () {
```
Comment (mchacki) [inline tests/js/common/shell/shell-cluster-collection.js]: I like the new tests, but why have the ShardKey tests been removed?

Comment (jsteemann) [reply tests/js/common/shell/shell-cluster-collection.js]: These tests assumed failures when using something like `_rev`, `_from`, `_to` as shard keys.
But they didn't work. The problem is the tests were written like this:
```
try {
  db._create(...);
} catch (err) {
  assert(...)
}
```
The problem is that the create _worked_ and the catch block was never reached.
So I added a `fail()` after the `db._create(...)` calls to fix this.
And then I found that all the creates worked... So it is possible to use these attributes as shard keys. Probably the tests were remainders from pre-3.0, when something like `_from` and `_to` were so special that they didn't work in many cases.

Comment (mchacki) [reply tests/js/common/shell/shell-cluster-collection.js]: Do we allow to shard by `_rev` really?
This value is changed on update.
We cannot change shard keys
=> We can create a collection that disallows upgrades?

Comment (mchacki) [reply tests/js/common/shell/shell-cluster-collection.js]: Ah i see _rev is silently ignored.
I am good with this PR

**Comment by KVS85** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/6497/

**Pull request merged**